src/overlays/actors/ovl_En_Knight/z_en_knight.c

```diff
@@ -16,184 +20,355 @@ void EnKnight_Destroy(Actor* thisx, PlayState* play);
+void EnKnight_KnightCaptainsHatCS(EnKnight* this, PlayState* play);
+void EnKnight_CaptainsHatCS(EnKnight* this, PlayState* play);
+void EnKnight_WaitIntroCutscene(EnKnight* this, PlayState* play);
+void EnKnight_IntroCutscene(EnKnight* this, PlayState* play);
+void EnKnight_SetupFlyingHead(EnKnight* this, PlayState* play);
+void EnKnight_FlyingHead(EnKnight* this, PlayState* play);
+void EnKnight_SetupFlyingHeadAttack(EnKnight* this, PlayState* play);
+void EnKnight_FlyingHeadAttack(EnKnight* this, PlayState* play);
+
+void EnKnight_UpdateEffects(EnKnight* this, PlayState* play);
+void EnKnight_DrawEffects(EnKnight* this, PlayState* play);
+
+void EnKnight_SetupTelegraphHeavyAttack(EnKnight* this, PlayState* play, s32 noTelegraph);
+void EnKnight_SetupStrafe(EnKnight* this, PlayState* play, s16 strafeAngle);
+void EnKnight_SetupRetreat(EnKnight* this, PlayState* play, u8 shielding);
+
+typedef struct {
+    /* 0x00 */ Vec3f pos;
+    /* 0x0C */ Vec3f velocity;
+    /* 0x18 */ Vec3f acceleration;
+    /* 0x24 */ UNK_TYPE1 unk_24[0x6];
+    /* 0x2A */ u8 active;
+    /* 0x2C */ s16 scrollTimer;
+    /* 0x2E */ s16 alpha;
+    /* 0x30 */ s16 unk_30; // set to 0 but unused
+    /* 0x34 */ f32 scale;
+    /* 0x38 */ f32 scaleTarget;
+} EnKnightEffect; // size = 0x3C
+
+static Vec3f sZeroVec = { 0.0f, 0.0f, 0.0f };
+
+u8 sSpawnIndex = 0;
+
+u8 sMinibossStartBgmCountdown = 0;
+
+typedef enum {
+    // None
+    KNIGHT_DMGEFF_STATE_0 = 0,
+
+    // Fire
+    KNIGHT_DMGEFF_STATE_1,
+    KNIGHT_DMGEFF_STATE_2,
+
+    // Ice
+    KNIGHT_DMGEFF_STATE_10 = 10,
+    KNIGHT_DMGEFF_STATE_11,
+
+    // Light
+    KNIGHT_DMGEFF_STATE_20 = 20,
+    KNIGHT_DMGEFF_STATE_21,
+
+    // Light again?
+    KNIGHT_DMGEFF_STATE_30 = 30,
+    KNIGHT_DMGEFF_STATE_31,
+
+    // Zora Barrier?
+    KNIGHT_DMGEFF_STATE_40 = 40,
+    KNIGHT_DMGEFF_STATE_41
+} KnightDamageEffectState;
+
+typedef enum {
+    /*  0 */ KNIGHT_DMGEFF_NONE,
+    /*  1 */ KNIGHT_DMGEFF_STUN,
+    /*  2 */ KNIGHT_DMGEFF_FIRE,
+    /*  3 */ KNIGHT_DMGEFF_ICE,
+    /*  4 */ KNIGHT_DMGEFF_LIGHT,
+    /* 10 */ KNIGHT_DMGEFF_ZORA_BARRIER = 10,
+    /* 13 */ KNIGHT_DMGEFF_LIGHT_RAY_FALLEN = 13,
+    /* 14 */ KNIGHT_DMGEFF_LIGHT_RAY,
+    /* 15 */ KNIGHT_DMGEFF_OTHER
+} EnKnightDamageEffect;
+
+static DamageTable sDamageTableStanding = {
+    /* Deku Nut       */ DMG_ENTRY(0, KNIGHT_DMGEFF_STUN),
+    /* Deku Stick     */ DMG_ENTRY(1, KNIGHT_DMGEFF_OTHER),
+    /* Horse trample  */ DMG_ENTRY(0, KNIGHT_DMGEFF_NONE),
+    /* Explosives     */ DMG_ENTRY(1, KNIGHT_DMGEFF_OTHER),
+    /* Zora boomerang */ DMG_ENTRY(0, KNIGHT_DMGEFF_STUN),
+    /* Normal arrow   */ DMG_ENTRY(3, KNIGHT_DMGEFF_OTHER),
+    /* UNK_DMG_0x06   */ DMG_ENTRY(0, KNIGHT_DMGEFF_NONE),
+    /* Hookshot       */ DMG_ENTRY(0, KNIGHT_DMGEFF_OTHER),
+    /* Goron punch    */ DMG_ENTRY(1, KNIGHT_DMGEFF_OTHER),
+    /* Sword          */ DMG_ENTRY(1, KNIGHT_DMGEFF_OTHER),
+    /* Goron pound    */ DMG_ENTRY(0, KNIGHT_DMGEFF_LIGHT_RAY_FALLEN),
```

Comment:
I kind of think we should name the damage effect based on the Goron Pound rather then the Light Ray Fallen since the fallen over damage function actually doesn't use the effects, so it doesn't really have meaning for the fallen over state.